### PR TITLE
webcore/Body: reject body mixin consumers with stored error instead of leaking it

### PR DESCRIPTION
**Pull request by robobun**
## Problem

When a `Response`/`Request` body has transitioned to the `.Error` state — e.g. the fetch body download fails after headers were received, an `HTMLRewriter` transform errors, or a `Bun.serve` request is aborted mid-upload — calling `.text()`/`.json()`/`.arrayBuffer()`/`.bytes()`/`.blob()`/`.formData()` fell through the `.Used`/`.Locked` checks in the body mixin and landed in `useAsAnyBlob()` / `useAsAnyBlobAllowNonUTF8String()`. There the `else =>` arm matched `.Error`, returned an empty `Blob`, and then unconditionally did:

```zig
this.* = .{ .Used = {} };
```

overwriting the `.Error` payload without calling `ValueError.deinit()`. That payload is one of:
- `jsc.Strong.Optional` (abort reason) — leaks a permanent GC root
- `jsc.SystemError` (network error with cloned URL) — leaks ref-counted struct + `bun.String`
- `bun.String` (HTMLRewriter message) — leaks a WTF string ref

In addition to the leak, the consumer silently resolved with an empty body instead of rejecting with the error, hiding the failure from user code.

The same overwrite-without-deinit also applied to the `.Locked` arm when `toAnyBlobAllowPromise()` returned `null` with a live `ReadableStream.Strong`.

## Repro

```js
const net = require("net");
let sock;
const server = net.createServer(s => {
  sock = s;
  s.write("HTTP/1.1 200 OK\\r\\nContent-Length: 100000\\r\\n\\r\\nhello");
});
server.listen(0, "127.0.0.1");
await once(server, "listening");

const res = await fetch(`http://127.0.0.1:${server.address().port}/`);
sock.destroy();                               // body download fails
for (let i = 0; i < 50; i++) await new Promise(r => setImmediate(r));

console.log(await res.text());               // before: "" (+ leaked SystemError)
                                              // after:  rejects ECONNRESET
```

## Fix

1. Each body mixin getter (`getText`, `getJSON`, `getArrayBuffer`, `getBytes`, `getFormData`, `getBlobWithThisValue`) now checks for `.Error` and rejects the returned promise with the stored error via `handleBodyError()`, which also releases the `ValueError` and marks the body `.Used`.
2. `useAsAnyBlob()` and `useAsAnyBlobAllowNonUTF8String()` now explicitly handle `.Error` (deinit before overwrite) and release the `.Locked` readable Strong when `toAnyBlobAllowPromise()` returns `null`, so any other caller of these helpers does not leak either.

## Verification

`test/js/web/fetch/body-mixin-errors.test.ts` adds a `consuming an errored body` suite that drives each body consumer through the `.Error` state and asserts it rejects with `ECONNRESET` (and that a second consume rejects with `ERR_BODY_ALREADY_USED`).

| | `text` | `json` | `arrayBuffer` | `bytes` | `blob` | `formData` |
|---|---|---|---|---|---|---|
| before | resolves `""` | `SyntaxError` | resolves `ArrayBuffer(0)` | resolves `Uint8Array(0)` | resolves empty `Blob` | `ERR_FORMDATA_PARSE_ERROR` |
| after | rejects `ECONNRESET` | rejects `ECONNRESET` | rejects `ECONNRESET` | rejects `ECONNRESET` | rejects `ECONNRESET` | rejects `ECONNRESET` |

Gate: with `src/` stashed the 7 new tests fail; with the fix they pass.

### CODE REVIEW

File: src/runtime/webcore/Body.zig [hunk 1]
```diff
@@ -1237,6 +1273,22 @@ pub fn Mixin(comptime Type: type) type {
...
+        /// `jsc.Strong`/`bun.String`/`SystemError` and silently returning an
+        /// empty body instead of surfacing the failure.
+        fn handleBodyError(value: *Body.Value, globalObject: *jsc.JSGlobalObject) JSValue {
+            var err = value.Error;
+            value.* = .{ .Used = {} };
+            const js_err = err.toJS(globalObject);
+            js_err.ensureStillAlive();
+            err.deinit();
+            return jsc.JSPromise.rejectedPromise(globalObject, js_err).toJS();
+        }
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/runtime/webcore/Body.zig:1306]: Not a leak — the `to*ErrorInstance()` methods consume their input ref:

```zig
// src/bun.js/bindings/SystemError.zig
pub fn toErrorInstance(this: *const SystemError, global: *JSGlobalObject) JSValue {
    defer this.deref();
    return SystemError__toErrorInstance(this, global);
}

// src/string.zig
pub fn toErrorInstance(this: *const String, globalObject: *jsc.JSGlobalObject) jsc.JSValue {
    defer this.deref();
    return JSC__createError(globalObject, this);
}
pub fn toTypeErrorInstance(this: *const String, globalObject: *jsc.JSGlobalObject) jsc.JSValue {
    defer this.deref();
    return JSC__createTypeError(globalObject, this);
}
```

So `ValueError.toJS()` transfers ownership: the `.SystemError`/`.Message`/`.TypeError` payload is consumed by `toErrorInstance()`, and the union is overwritten with a fresh `.JSValue` Strong pointing at the resulting JS error. `err.deinit()` then releases that Strong. `.AbortReason` is a plain enum with no refcount. This is the same contract the existing `this.Error.toJS(global)` call sites in `toErrorInstance()` rely on.

File: src/http/h3_client/AltSvc.zig [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+};
+
+var cache: bun.StringHashMapUnmanaged(Record) = .{};
+
+/// Hard cap on cached origins. When reached, `record()` first sweeps expired
+/// entries and then refuses the new insert if still full — bounded memory for
+/// long-lived processes that hit many distinct origins.
+const max_entries = 256;
+
+fn key(buf: []u8, hostname: []const u8, port: u16) []const u8 {
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/http/h3_client/AltSvc.zig:90]: These files are not part of this PR — they were accidentally staged in an intermediate commit (`c2779c2`) that has since been force-pushed over. The current branch head (`852f782`) only touches `src/bun.js/webcore/Body.zig` and `test/js/web/fetch/body-mixin-errors.test.ts`.

File: src/http/h3_client/AltSvc.zig [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+/// Look up a previously-advertised h3 alternative for `origin_host:origin_port`.
+/// Expired entries are dropped on access. Runs on the HTTP thread inside
+/// `start_()`.
+pub fn lookup(origin_host: []const u8, origin_port: u16) ?u16 {
+    var buf: [256 + 8]u8 = undefined;
+    if (origin_host.len > 256) return null;
+    const k = key(&buf, origin_host, origin_port);
+    const rec = cache.get(k) orelse return null;
+    if (std.time.timestamp() >= rec.expires_at) {
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/http/h3_client/AltSvc.zig:148]: These files are not part of this PR — they were accidentally staged in an intermediate commit (`c2779c2`) that has since been force-pushed over. The current branch head (`852f782`) only touches `src/bun.js/webcore/Body.zig` and `test/js/web/fetch/body-mixin-errors.test.ts`.

File: src/http/h3_client/ClientSession.zig [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+        // the stream instead of leaking it on the pooled session.
+        if (!stream.request_body_done) qs.reset();
+    }
+    stream.qstream = null;
+    if (std.mem.indexOfScalar(*Stream, this.pending.items, stream)) |i| {
+        _ = this.pending.orderedRemove(i);
+    }
+    stream.deinit();
+    this.deref();
+}
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/http/h3_client/ClientSession.zig:99]: These files are not part of this PR — they were accidentally staged in an intermediate commit (`c2779c2`) that has since been force-pushed over. The current branch head (`852f782`) only touches `src/bun.js/webcore/Body.zig` and `test/js/web/fetch/body-mixin-errors.test.ts`.

File: src/http/h3_client/encode.zig [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+            qs.shutdown();
+            client.state.request_stage = .done;
+        } else if (!drained) {
+            qs.wantWrite(true);
+        } else if (data.len > 0) {
+            sb.reportDrain();
+        }
+        sb.release();
+        if (stream.request_body_done) body.detach();
+        return;
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/http/h3_client/encode.zig:108]: These files are not part of this PR — they were accidentally staged in an intermediate commit (`c2779c2`) that has since been force-pushed over. The current branch head (`852f782`) only touches `src/bun.js/webcore/Body.zig` and `test/js/web/fetch/body-mixin-errors.test.ts`.

File: test/js/web/fetch/body-mixin-errors.test.ts [hunk 6]
```diff
@@ -21,4 +23,93 @@ describe("body-mixin-errors", () => {
...
+    // `useAsAnyBlob*()` which overwrote it with `.Used` without calling
+    // `ValueError.deinit()`, leaking the payload and silently returning an
+    // empty body.
+    async function withErroredBody(fn: (res: Response) => Promise<unknown>) {
+      let currentSocket: Socket | undefined;
+      const server = createServer(socket => {
+        currentSocket = socket;
+        // Promise a large body but only send a few bytes so that closing the
+        // socket produces a body download failure.
+        socket.write("HTTP/1.1 200 OK\r\nContent-Length: 100000\r\n\r\nhello");
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/web/fetch/body-mixin-errors.test.ts:42]: Intentionally excluded. The `.Error` branch in `getFormData` is the same `handleBodyError(value, globalObject)` call added to the other five getters, so its behaviour is already exercised by the text/json/arrayBuffer/bytes/blob cases.

Adding a `Content-Type` header here would only change the `.Locked` fallback path (the one this test is *not* about): with `application/x-www-form-urlencoded`, `getFormDataEncoding()` succeeds and `setPromise(.getFormData, ...)` reads the body stream, which then rejects via `readableStreamToFormData` rather than `handleBodyError`. That is a different code path with a different error shape, and under load the body can still be in `.Locked` when the consumer runs — so the assertion would have to accept two different errors, making the test less precise, not more. The comment above the `it.each` documents this.

File: test/js/web/fetch/body-mixin-errors.test.ts [hunk 7]
```diff
@@ -21,4 +23,93 @@ describe("body-mixin-errors", () => {
...
+        // the body transitions from `.Locked` to `.Error`. `Bun.sleep` (as
+        // opposed to `setImmediate`) actually yields the CPU so the HTTP
+        // thread can run even when the JS thread would otherwise be busy
+        // under load.
+        const closed = once(currentSocket!, "close");
+        currentSocket!.destroy();
+        await closed;
+        for (let i = 0; i < 20; i++) {
+          await Bun.sleep(1);
+        }
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/web/fetch/body-mixin-errors.test.ts:64]: This is deliberate and was the fix for the previous gate failure. The condition being awaited — "the HTTP client thread has observed the socket close and posted the body failure to the JS event loop" — has no JS-observable signal, so it cannot be polled.

The first iteration of this test used 50× `setImmediate` (pure macrotask barrier, equivalent to `Bun.sleep(0)`) and was **flaky under load on the release build**: 15 concurrent processes pinned to one core produced failures because `setImmediate` only yields to *other JS tasks on the same thread* — the HTTP client thread never got scheduled, so the body stayed in `.Locked` and `formData()` / the re-consume test hit the wrong path.

`Bun.sleep(1)` goes through the timer queue and actually yields the CPU to the OS scheduler for ≥1 ms, which lets the HTTP thread run even when the JS thread is saturated. Verified under the same 15-process/1-core stress: 0 failures across 30+ runs on both release and ASAN. The total wait is bounded at ~20 ms, which is well under any test timeout. The comment above the loop explains this; see also the precedent in `test/js/web/fetch/fetch-http2-client.test.ts` ("Content-Length satisfied before END_STREAM") which uses `Bun.sleep(80)` for the same cross-thread-synchronisation reason.
